### PR TITLE
go/common: Change Namespace's and Hash's text (un)marshaling to use the hex form

### LINKED ISSUE #4078: Use hex-encoded runtime IDs and runtime state roots everywhere
##### SUMMARY
Currently, running `oasis-node control status` contains Base64-encoded runtime IDs and state roots:

```

  ... trimmed ...

  "runtimes": {
    "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=": {
      "descriptor": {
        "v": 2,
        "id": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=",
        "entity_id": "4eM2SdjntkGGH0DqhloG8HJsPNSoEeYHFACuOV+9qbc=",
        "genesis": {
          "state_root": "xnK40e9W7Sirh8NiLFEUBpvdOte4+XN0mNDAHs7wlno=",
          "state": null,
          "storage_receipts": null,
          "round": 0
        },

  ... trimmed ...

```

but other things (docs, CLI) contain hex-encoded versions.

Add custom JSON encoder/decoder for runtime IDs and runtime state root (anything else?) that would encode them to hex instead of Base64 for more consistent UX with CLI parameters.

##### ISSUE TYPE
- Feature Idea

### PR DESCRIPTION
**Pull request by tjanez**
Changing `Namespace`'s and `Hash`'s text (un)marshaling from the current Base64-encoded form to hex-encoded form will unify the textual representation of runtime IDs and runtime state roots with Docs and CLI flags.

Closes #4078.

Example of previous `"runtimes"` output in `oasis-node control status`:

```json
  "runtimes": {
    "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAE=": {
      "descriptor": {
        "v": 2,
        "id": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAE=",
        "entity_id": "Ldzg8aeLiUBrMYxidd5DqEzpamyV2cprmRH0pG8d/Jg=",
        "genesis": {
          "state_root": "xnK40e9W7Sirh8NiLFEUBpvdOte4+XN0mNDAHs7wlno=",
          "state": null,
          "storage_receipts": null,
          "round": 0
        },
        "kind": 1,
        "tee_hardware": 0,
        "versions": {
          "version": {
            "minor": 2
          }
        },
        "executor": {
          "group_size": 3,
          "group_backup_size": 3,
          "allowed_stragglers": 1,
          "round_timeout": 5,
          "max_messages": 256
        },
        "txn_scheduler": {
          "algorithm": "simple",
          "batch_flush_timeout": 1000000000,
          "max_batch_size": 100,
          "max_batch_size_bytes": 1048576,
          "propose_batch_timeout": 2
        },
        "storage": {
          "group_size": 3,
          "min_write_replication": 2,
          "max_apply_write_log_entries": 10000,
          "max_apply_ops": 2,
          "checkpoint_interval": 100,
          "checkpoint_num_kept": 2,
          "checkpoint_chunk_size": 8388608
        },
        "admission_policy": {
          "any_node": {}
        },
        "constraints": {
          "executor": {
            "backup-worker": {
              "max_nodes": {
                "limit": 1
              },
              "min_pool_size": {
                "limit": 3
              }
            },
            "worker": {
              "max_nodes": {
                "limit": 1
              },
              "min_pool_size": {
                "limit": 3
              }
            }
          },
          "storage": {
            "worker": {
              "max_nodes": {
                "limit": 1
              },
              "min_pool_size": {
                "limit": 3
              }
            }
          }
        },
        "staking": {},
        "governance_model": "entity"
      },
      "latest_round": 947,
      "latest_hash": "z9h9PdT5KWCMGGHh8Zini+0y0wyEtRl68dXuyZbLbac=",
      "latest_time": "2021-09-23T12:04:43Z",
      "latest_state_root": {
        "ns": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=",
        "version": 947,
        "root_type": 1,
        "hash": "3arshA5pbJZxljHGtrWrEVLnMo2Nkhyo0/P6SFVuOF4="
      },
      "genesis_round": 0,
      "genesis_hash": "rtlMA+vS0W37X2Q0Ahq/acjBX8abaxlVTSPailoFN3Y=",
      "committee": {
        "latest_round": 947,
        "latest_height": 5939575,
        "last_committee_update_height": 5939570,
        "executor_roles": [
          "worker",
          "backup-worker"
        ],
        "storage_roles": [
          "worker"
        ],
        "is_txn_scheduler": true,
        "peers": [
          "/ip4/12.71.39.73/tcp/41002/p2p/12D3KooWJvL8mYzHbcLtj91bf5sHhtrB7C8CWND5sV6Kk24eUdpQ",
          "/ip4/106.67.32.45/tcp/26648/p2p/12D3KooWBKgcH7TGMSLuxzLxK41nTwk6DsxHRpb7HpWQXJzLurcv"
        ]
      },
      "storage": {
        "last_finalized_round": 947
      }
    }
  }
```

Example of new `"runtimes"` output in `oasis-node control status`:

```json
  "runtimes": {
    "0000000000000000000000000000000000000000000000000000000000000001": {
      "descriptor": {
        "v": 2,
        "id": "0000000000000000000000000000000000000000000000000000000000000001",
        "entity_id": "Ldzg8aeLiUBrMYxidd5DqEzpamyV2cprmRH0pG8d/Jg=",
        "genesis": {
          "state_root": "c672b8d1ef56ed28ab87c3622c5114069bdd3ad7b8f9737498d0c01ecef0967a",
          "state": null,
          "storage_receipts": null,
          "round": 0
        },
        "kind": 1,
        "tee_hardware": 0,
        "versions": {
          "version": {
            "minor": 2
          }
        },
        "executor": {
          "group_size": 3,
          "group_backup_size": 3,
          "allowed_stragglers": 1,
          "round_timeout": 5,
          "max_messages": 256
        },
        "txn_scheduler": {
          "algorithm": "simple",
          "batch_flush_timeout": 1000000000,
          "max_batch_size": 100,
          "max_batch_size_bytes": 1048576,
          "propose_batch_timeout": 2
        },
        "storage": {
          "group_size": 3,
          "min_write_replication": 2,
          "max_apply_write_log_entries": 10000,
          "max_apply_ops": 2,
          "checkpoint_interval": 100,
          "checkpoint_num_kept": 2,
          "checkpoint_chunk_size": 8388608
        },
        "admission_policy": {
          "any_node": {}
        },
        "constraints": {
          "executor": {
            "backup-worker": {
              "max_nodes": {
                "limit": 1
              },
              "min_pool_size": {
                "limit": 3
              }
            },
            "worker": {
              "max_nodes": {
                "limit": 1
              },
              "min_pool_size": {
                "limit": 3
              }
            }
          },
          "storage": {
            "worker": {
              "max_nodes": {
                "limit": 1
              },
              "min_pool_size": {
                "limit": 3
              }
            }
          }
        },
        "staking": {},
        "governance_model": "entity"
      },
      "latest_round": 947,
      "latest_hash": "cfd87d3dd4f929608c1861e1f198a78bed32d30c84b5197af1d5eec996cb6da7",
      "latest_time": "2021-09-23T14:04:43+02:00",
      "latest_state_root": {
        "ns": "0000000000000000000000000000000000000000000000000000000000000000",
        "version": 947,
        "root_type": 1,
        "hash": "ddaaec840e696c96719631c6b6b5ab1152e7328d8d921ca8d3f3fa48556e385e"
      },
      "genesis_round": 0,
      "genesis_hash": "aed94c03ebd2d16dfb5f6434021abf69c8c15fc69b6b19554d23da8a5a053776",
      "committee": {
        "latest_round": 947,
        "latest_height": 5939575,
        "last_committee_update_height": 5939570,
        "executor_roles": [
          "worker",
          "backup-worker"
        ],
        "storage_roles": [
          "worker"
        ],
        "is_txn_scheduler": true,
        "peers": [
          "/ip4/12.71.39.73/tcp/41002/p2p/12D3KooWJvL8mYzHbcLtj91bf5sHhtrB7C8CWND5sV6Kk24eUdpQ",
          "/ip4/106.67.32.45/tcp/26648/p2p/12D3KooWBKgcH7TGMSLuxzLxK41nTwk6DsxHRpb7HpWQXJzLurcv"
        ]
      },
      "storage": {
        "last_finalized_round": 947
      }
    }
  },

```

### CODE REVIEW

**Review by pro-wh — COMMENTED** [review]
implementation looks fine, and I agree that hex is easier.

but I wonder if this same logic could have applied to other short binary strings too, such as entity IDs and other public keys. somehow we came out of that discussion with the idea to make everything base64 where possible. what had we reasoned back then, and how does it relate to runtime namespaces and Hash?

**Comment by kostko** [comment]
One problem with this is also that it probably prevents loading the current genesis files.

**Comment by tjanez** [comment]
> One problem with this is also that it probably prevents loading the current genesis files.

Yes, that is correct. Loading the [current Mainnet genesis file](https://github.com/oasisprotocol/mainnet-artifacts/releases/download/2021-04-28/genesis.json) would fail with:
```
{"caller":"node.go:673","err":"initGenesis: failed to create local genesis file provider: genesis: malformed genesis file: encoding/hex: invalid byte: U+0047 'G'","level":"error","module":"oasis-node","msg":"failed to initialize the genesis provider","ts":"2021-09-23T15:13:08.313232561Z"}
```

I've changed `Namespace`'s and `Hash`'s `UnmarshalText()` implementation to first try decoding it as a hex-encoded string and then fallback to accepting the current Base64-encoded form.

This should enable loading the current Mainnet genesis file with this PR.

Running `oasis-node genesis check -g genesis.json` should return that the genesis file is not in canonical form:
```
ts=2021-09-23T20:25:20.14847867Z level=error module=cmd/genesis caller=genesis.go:696 msg="genesis file is not in canonical form" err="genesis file is not in canonical form, see the diff on stderr"
Diff:
--- Actual
+++ Canonical
@@ -892,10 +892,10 @@
     "runtimes": [
       {
         "v": 2,
-        "id": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA/wM=",
+        "id": "000000000000000000000000000000000000000000000000000000000000ff03",
         "entity_id": "RDCyARTIdpiG75L9vVXAYIhRCToSrchZkv3JUW3Pji8=",
         "genesis": {
-          "state_root": "JzyPF+MZtIkjNUTAWixK6BcWaAHjUpFkQGtVarmdd1I=",
+          "state_root": "273c8f17e319b489233544c05a2c4ae817166801e3529164406b556ab99d7752",
           "state": null,
           "storage_receipts": null,
           "round": 43825
@@ -965,10 +965,10 @@
     "suspended_runtimes": [
       {
         "v": 2,
-        "id": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA/wE=",
+        "id": "000000000000000000000000000000000000000000000000000000000000ff01",
         "entity_id": "RDCyARTIdpiG75L9vVXAYIhRCToSrchZkv3JUW3Pji8=",
         "genesis": {
-          "state_root": "jHjpzasZlD2QcX48fUfJOfcathW/tBO7qFi3df7iKsw=",
+          "state_root": "8c78e9cdab19943d90717e3c7d47c939f71ab615bfb413bba858b775fee22acc",
           "state": [
             [
               "Z2VuZXNpc19pbml0aWFsaXplZA==",
@@ -1100,10 +1100,10 @@
       },
       {
         "v": 2,
-        "id": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA/wA=",
+        "id": "000000000000000000000000000000000000000000000000000000000000ff00",
         "entity_id": "RDCyARTIdpiG75L9vVXAYIhRCToSrchZkv3JUW3Pji8=",
         "genesis": {
-          "state_root": "9ygKjgFpjSrbjrCrXgS/UHEVs8HujW+lmJBPAxBpklY=",
+          "state_root": "f7280a8e01698d2adb8eb0ab5e04bf507115b3c1ee8d6fa598904f0310699256",
           "state": [
             [
               "HMooYA10kTZVILMbRm+IZHuYOew=",
@@ -1182,10 +1182,10 @@
       },
       {
         "v": 2,
-        "id": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA/wI=",
+        "id": "000000000000000000000000000000000000000000000000000000000000ff02",
         "entity_id": "RDCyARTIdpiG75L9vVXAYIhRCToSrchZkv3JUW3Pji8=",
         "genesis": {
-          "state_root": "cQJrk1FPGBLzGEaiR6JE1rVRl7RlqPttbEMmMy0uW/A=",
+          "state_root": "71026b93514f1812f31846a247a244d6b55197b465a8fb6d6c4326332d2e5bf0",
           "state": [
             [
               "IqoJboluFsC5m9uT1ghMlHw73as=",
@@ -4177,26 +4177,26 @@
       "max_evidence_age": 100
     },
     "runtime_states": {
-      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA/wA=": {
-        "state_root": "Jyht8nNAlEznxg0JvKLbCB0ZVos5uALw8caG3GQIsUo=",
+      "000000000000000000000000000000000000000000000000000000000000ff00": {
+        "state_root": "27286df27340944ce7c60d09bca2db081d19568b39b802f0f1c686dc6408b14a",
         "state": [],
         "storage_receipts": [],
         "round": 53355
       },
-      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA/wE=": {
-        "state_root": "xaJaKDXXM2w2TYMWg0KMuAXuWjRgNDBAFCZCzCuo8KI=",
+      "000000000000000000000000000000000000000000000000000000000000ff01": {
+        "state_root": "c5a25a2835d7336c364d831683428cb805ee5a3460343040142642cc2ba8f0a2",
         "state": [],
         "storage_receipts": [],
         "round": 2269
       },
-      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA/wI=": {
-        "state_root": "cQJrk1FPGBLzGEaiR6JE1rVRl7RlqPttbEMmMy0uW/A=",
+      "000000000000000000000000000000000000000000000000000000000000ff02": {
+        "state_root": "71026b93514f1812f31846a247a244d6b55197b465a8fb6d6c4326332d2e5bf0",
         "state": [],
         "storage_receipts": [],
         "round": 1
       },
-      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA/wM=": {
-        "state_root": "JzyPF+MZtIkjNUTAWixK6BcWaAHjUpFkQGtVarmdd1I=",
+      "000000000000000000000000000000000000000000000000000000000000ff03": {
+        "state_root": "273c8f17e319b489233544c05a2c4ae817166801e3529164406b556ab99d7752",
         "state": [],
         "storage_receipts": [],
         "round": 43825

```
But otherwise, the canonical file is valid and can be loaded correctly.

Running ` oasis-node-dev debug fix-genesis -g genesis.json --genesis.new_file genesis-new.json` will convert it to the new canonical form with hex-encoded runtime IDs and runtime state roots:

```
$ oasis-node genesis check -g genesis-new.json
genesis file is valid and in canonical form
genesis document's hash: 53852332637bacb61b91b6411ab4095168ba02a50be4c3f82448438826f23898
genesis file's SHA256 checksum: 180eab5177f5fbd4e7e63d99eb496c0e2e127ec7bc8104ddfa342e6a9870e06e
```

Trying to load the new genesis file in the new canonical form would result in an error:
```
$ oasis-node genesis check -g genesis-new.json
ts=2021-09-23T20:25:52.134173612Z level=error module=cmd/genesis caller=genesis.go:662 msg="failed to open genesis file" err="genesis: malformed genesis file: malformed namespace"
```
which is what we want.

**Comment by tjanez** [comment]
> but I wonder if this same logic could have applied to other short binary strings too, such as entity IDs and other public keys. somehow we came out of that discussion with the idea to make everything base64 where possible. what had we reasoned back then, and how does it relate to runtime namespaces and Hash?

As far as I remember, we had a discrepancy between what users needed to pass as CLI arguments (hex-encoded values) and how CLI printed the values (Base64-encoded) values (see https://github.com/oasisprotocol/oasis-core/issues/2588) and decided to use Base64-encoding for various IDs (i.e. entity ID, node ID, runtime ID).

I think the advantages of using Base64-encoding for the entity and node IDs are that:
- They are shorter and take less space when they are shown in block explorers.
- Pem files containing public keys use Base64-encoding so it is very straightforward to obtain one's entity or node ID.
- When an entity or an entity transaction needs to be signed on the Ledger hardware wallet, using Base64-encoding enables showing the whole entity ID in a single (or at most 2) line(s).

For runtime IDs, I think the practice has always been to reason with them in hex-encoded form. Examples are:
- `RUNTIME_ID` in https://github.com/oasisprotocol/oasis-core/blob/640adb6a8b24da33c32ec2a8edb28fcb18c4c217/docs/setup/deploying-a-runtime.md#initializing-a-runtime.
- Runtime CLI flags and/or YAML configuration uses hex-encoded runtime IDs:
```
# Runtimes.
runtime:
  # List of supported Runtime IDs.
  supported:
    - "0000000000000000000000000000000000000000000000000000000000000001"
  # List of Runtime binaries of the form: <RUNTIME-ID>: <RUNTIME-BINARY-PATH>
  paths:
    "0000000000000000000000000000000000000000000000000000000000000001": "/srv/oasis/runtimes//cipher-paratime"
```

And for `Hash` (e.g. state roots), it was outputted as an array of bytes instead of a string (see https://github.com/oasisprotocol/oasis-core/issues/2476) and we decided to go with a Base64-encoded strings back then.

But also here, I think the practice has been to reason with them in hex-encoded form. Examples are:
- Output of `oasis-node storage migrate` uses hex-encoded state roots:
```
...
- migrating from v4 to v5...
- migrating version 24468...
- migrated root state-root:195cf7a9a103e7300b2bb4e537cb9935cbebd83e448e67aa55433861a6ad7426 -> state-root:cea105a5d701deab935b94af9e8e0c5af5dcdb61c242bf434da9f11aa8d110ba
- migrated root io-root:0850c5a33ee7f45aa92724b7d5f28c9ac9ae8799b88cc5be9773e8aba9526ca7 -> io-root:19713a2b44e1bf868ebee43c36872baa3058870bb890a5e25d1c4cea2622be77
- migrated root io-root:477391131f60ac2c22bce9167c7e3783a13d4fb81fddd2d388b4ead6a586fe52 -> io-root:f29f86d491303c5fd7b3572e97cbd65b7487b6b4ac519623afd161cc2e4678b7
```

**Comment by pro-wh** [comment]
ok thanks for the info

File: go/oasis-test-runner/scenario/e2e/genesis_file.go [hunk 1]
```diff
@@ -93,15 +98,45 @@ func (s *genesisFileImpl) Run(childEnv *env.Env) error {
...
+	case !strings.Contains(err.Error(), expectedError):
+		return fmt.Errorf(
+			"e2e/genesis-file: running genesis check for the latest Mainnet genesis "+
+				"file should fail with an error containing: '%s' (actual error: %s)",
+			expectedError, err,
+		)
+	default:
+		s.Logger.Info("latest Mainnet genesis file is OK, but is not in canonical form")
+	}
+
```
Comment (kostko) [inline go/oasis-test-runner/scenario/e2e/genesis_file.go:125]: The test should probably also make sure that the genesis document hash matches?

Comment (tjanez) [reply go/oasis-test-runner/scenario/e2e/genesis_file.go:125]: Yes, good idea.

Comment (tjanez) [reply go/oasis-test-runner/scenario/e2e/genesis_file.go:128]: Done.

**Review by kostko — APPROVED** [review]

**Pull request merged**